File: V2/decipher/decipher.go
// Package decipher provides signature and n-parameter deciphering for YouTube URLs.
// This is required to bypass throttling and obtain valid stream URLs.
package decipher

import (
	"fmt"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"sync"
)

// Decipherer handles signature and n-parameter challenges
type Decipherer struct {
	playerCode    string
	playerURL     string

	sigTokens     []string
	nSolver       *NSolver

	mu sync.Mutex
}

// JavaScript regex patterns for signature function extraction
var (
	jsVarStr       = `[a-zA-Z_\$]\w*`
	jsSingleQuote  = `'[^'\\]*(?:\\[\s\S][^'\\]*)*'`
	jsDoubleQuote  = `"[^"\\]*(?:\\[\s\S][^"\\]*)*"`
	jsQuoteStr     = fmt.Sprintf(`(?:%s|%s)`, jsSingleQuote, jsDoubleQuote)
	jsKeyStr       = fmt.Sprintf(`(?:%s|%s)`, jsVarStr, jsQuoteStr)
	jsPropStr      = fmt.Sprintf(`(?:\.%s|\[%s\])`, jsVarStr, jsQuoteStr)

	reverseStr = `:function\(a\)\{(?:return )?a\.reverse\(\)\}`
	sliceStr   = `:function\(a,b\)\{return a\.slice\(b\)\}`
	spliceStr  = `:function\(a,b\)\{a\.splice\(0,b\)\}`
	swapStr    = `:function\(a,b\)\{var c=a\[0\];a\[0\]=a\[b(?:%a\.length)?\];a\[b(?:%a\.length)?\]=c(?:;return a)?\}`

	actionsObjRegex  = regexp.MustCompile(fmt.Sprintf(
		`var (%s)=\{((?:(?:%s%s|%s%s|%s%s|%s%s),?\r?\n?)+)\};`,
		jsVarStr, jsKeyStr, reverseStr, jsKeyStr, sliceStr, jsKeyStr, spliceStr, jsKeyStr, swapStr))

	actionsFuncRegex = regexp.MustCompile(fmt.Sprintf(
		`function(?: %s)?\(a\)\{a=a\.split\((?:''|"")\);\s*((?:(?:a=)?%s%s\(a,\d+\);)+)return a\.join\((?:''|"")\)\}`,
		jsVarStr, jsVarStr, jsPropStr))

	reverseRegex = regexp.MustCompile(fmt.Sprintf(`(?:^|,)(%s)%s`, jsKeyStr, reverseStr))
	sliceRegex   = regexp.MustCompile(fmt.Sprintf(`(?:^|,)(%s)%s`, jsKeyStr, sliceStr))
	spliceRegex  = regexp.MustCompile(fmt.Sprintf(`(?:^|,)(%s)%s`, jsKeyStr, spliceStr))
	swapRegex    = regexp.MustCompile(fmt.Sprintf(`(?:^|,)(%s)%s`, jsKeyStr, swapStr))

	// N-parameter function patterns
	nFuncNameRegex = regexp.MustCompile(`\.get\("n"\)\)&&\(b=([a-zA-Z0-9$]+)(?:\[(\d+)\])?\([a-zA-Z0-9]\)`)
	nFuncBodyRegex = regexp.MustCompile(`(?s)var %s=\{.*?\};`)
)

// New creates a new Decipherer with the given player JS code
func New(playerCode, playerURL string) (*Decipherer, error) {
	d := &Decipherer{
		playerCode: playerCode,
		playerURL:  playerURL,
	}

	if err := d.extractSignatureTokens(); err != nil {
		return nil, fmt.Errorf("failed to extract signature tokens: %w", err)
	}

	return d, nil
}

// NewDecipherer creates a new Decipherer from player code
func NewDecipherer(playerCode string) (*Decipherer, error) {
	d := &Decipherer{
		playerCode: playerCode,
	}

	if err := d.extractSignatureTokens(); err != nil {
		return nil, fmt.Errorf("failed to extract signature tokens: %w", err)
	}

	// Initialize n-solver
	nSolver, err := NewNSolver(playerCode)
	if err != nil {
		// N-solver is optional, some videos don't need it
		d.nSolver = nil
	} else {
		d.nSolver = nSolver
	}

	return d, nil
}

// GetSignatureTimestamp returns the signature timestamp from the player code
func (d *Decipherer) GetSignatureTimestamp() int {
	return GetSignatureTimestamp(d.playerCode)
}

// SolveNChallenge solves the n-parameter challenge using the JS runtime
func (d *Decipherer) SolveNChallenge(n string) (string, error) {
	if d.nSolver == nil {
		return n, nil
	}
	return d.nSolver.Solve(n)
}

// DecipherURL deciphers a stream URL by solving signature and n-parameter challenges
func (d *Decipherer) DecipherURL(streamURL string) (string, error) {
	parsed, err := url.Parse(streamURL)
	if err != nil {
		return "", fmt.Errorf("failed to parse URL: %w", err)
	}

	query := parsed.Query()

	// Check for signature cipher in query
	if sig := query.Get("s"); sig != "" {
		// Decipher signature
		decipheredSig := d.decipherSignature(sig)

		// Get signature parameter name (usually "sig" or "signature")
		sp := query.Get("sp")
		if sp == "" {
			sp = "signature"
		}

		query.Set(sp, decipheredSig)
		query.Del("s")
		query.Del("sp")
	}

	// Handle n-parameter for throttle bypass
	if n := query.Get("n"); n != "" {
		newN, err := d.solveNChallenge(n)
		if err == nil && newN != "" {
			query.Set("n", newN)
		}
	}

	parsed.RawQuery = query.Encode()
	return parsed.String(), nil
}

// DecipherSignature deciphers a signature using the extracted tokens
func (d *Decipherer) DecipherSignature(sig string) string {
	return d.decipherSignature(sig)
}

// decipherSignature applies the signature transformation
func (d *Decipherer) decipherSignature(sig string) string {
	arr := strings.Split(sig, "")

	for _, token := range d.sigTokens {
		if len(token) < 1 {
			continue
		}

		switch token[0] {
		case 'r':
			// Reverse
			for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
				arr[i], arr[j] = arr[j], arr[i]
			}

		case 's':
			// Slice
			if pos, err := strconv.Atoi(token[1:]); err == nil && pos < len(arr) {
				arr = arr[pos:]
			}

		case 'p':
			// Splice
			if pos, err := strconv.Atoi(token[1:]); err == nil && pos < len(arr) {
				arr = arr[pos:]
			}

		case 'w':
			// Swap
			if pos, err := strconv.Atoi(token[1:]); err == nil {
				swapPos := pos % len(arr)
				arr[0], arr[swapPos] = arr[swapPos], arr[0]
			}
		}
	}

	return strings.Join(arr, "")
}

// extractSignatureTokens extracts signature transformation tokens from player code
func (d *Decipherer) extractSignatureTokens() error {
	objects := actionsObjRegex.FindStringSubmatch(d.playerCode)
	functions := actionsFuncRegex.FindStringSubmatch(d.playerCode)

	if len(objects) < 3 || len(functions) < 2 {
		// Try alternative extraction methods
		return d.extractSignatureTokensAlt()
	}

	obj := strings.ReplaceAll(objects[1], "$", "\\$")
	objBody := strings.ReplaceAll(objects[2], "$", "\\$")
	funcBody := strings.ReplaceAll(functions[1], "$", "\\$")

	reverseKey := extractKey(reverseRegex, objBody)
	sliceKey := extractKey(sliceRegex, objBody)
	spliceKey := extractKey(spliceRegex, objBody)
	swapKey := extractKey(swapRegex, objBody)

	keys := fmt.Sprintf("(%s|%s|%s|%s)", reverseKey, sliceKey, spliceKey, swapKey)
	tokenizeRegex := regexp.MustCompile(fmt.Sprintf(
		`(?:a=)?%s(?:\.%s|\[(?:'%s'|"%s")\])\(a,(\d+)\)`,
		obj, keys, keys, keys))

	matches := tokenizeRegex.FindAllStringSubmatch(funcBody, -1)

	for _, result := range matches {
		if len(result) < 5 {
			continue
		}

		key := result[1]
		if key == "" {
			key = result[2]
		}
		if key == "" {
			key = result[3]
		}

		switch key {
		case reverseKey:
			d.sigTokens = append(d.sigTokens, "r")
		case sliceKey:
			d.sigTokens = append(d.sigTokens, "s"+result[4])
		case spliceKey:
			d.sigTokens = append(d.sigTokens, "p"+result[4])
		case swapKey:
			d.sigTokens = append(d.sigTokens, "w"+result[4])
		}
	}

	return nil
}

// extractSignatureTokensAlt tries alternative patterns for signature extraction
func (d *Decipherer) extractSignatureTokensAlt() error {
	// Alternative signature function pattern
	altPattern := regexp.MustCompile(`\b[cs]\s*&&\s*[adf]\.set\([^,]+\s*,\s*encodeURIComponent\(([a-zA-Z0-9$]+)\(`)
	match := altPattern.FindStringSubmatch(d.playerCode)

	if len(match) < 2 {
		// If no signature function found, the URL may not need deciphering
		return nil
	}

	return nil
}

// solveNChallenge solves the n-parameter challenge to bypass throttling
func (d *Decipherer) solveNChallenge(n string) (string, error) {
	// The n-parameter solving requires JavaScript execution
	// This is a simplified version that may need enhancement

	// For now, we'll return the original n value
	// Full implementation would use a JS runtime like goja
	return n, nil
}

// extractKey extracts a key from the object body using the given regex
func extractKey(re *regexp.Regexp, body string) string {
	match := re.FindStringSubmatch(body)
	if len(match) < 2 {
		return ""
	}

	key := match[1]
	// Remove quotes if present
	key = strings.Trim(key, `"'`)
	return key
}

// GetSignatureTimestamp extracts the signature timestamp from player code
func GetSignatureTimestamp(playerCode string) int {
	patterns := []string{
		`(?:signatureTimestamp|sts)\s*:\s*(\d{5})`,
		`"STS"\s*:\s*(\d{5})`,
	}

	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		match := re.FindStringSubmatch(playerCode)
		if len(match) >= 2 {
			if sts, err := strconv.Atoi(match[1]); err == nil {
				return sts
			}
		}
	}

	return 0
}

// ExtractPlayerID extracts the player ID from a player URL
func ExtractPlayerID(playerURL string) string {
	patterns := []string{
		`/s/player/([a-zA-Z0-9_-]{8,})/`,
		`/([a-zA-Z0-9_-]{8,})/player`,
		`\b(vfl[a-zA-Z0-9_-]+)\b`,
	}

	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		match := re.FindStringSubmatch(playerURL)
		if len(match) >= 2 {
			return match[1]
		}
	}

	return ""
}

File: TS/src/decipher/decipher.ts
// Package decipher provides signature and n-parameter deciphering for YouTube URLs.
// This is required to bypass throttling and obtain valid stream URLs.

import { NSolver } from "./nsolver.js";

// JavaScript regex patterns for signature function extraction
const jsVarStr = `[a-zA-Z_\\$]\\w*`;
const jsSingleQuote = `'[^'\\\\]*(?:\\\\[\\s\\S][^'\\\\]*)*'`;
const jsDoubleQuote = `"[^"\\\\]*(?:\\\\[\\s\\S][^"\\\\]*)*"`;
const jsQuoteStr = `(?:${jsSingleQuote}|${jsDoubleQuote})`;
const jsKeyStr = `(?:${jsVarStr}|${jsQuoteStr})`;
const jsPropStr = `(?:\\.${jsVarStr}|\\[${jsQuoteStr}\\])`;

const reverseStr = `:function\\(a\\)\\{(?:return )?a\\.reverse\\(\\)\\}`;
const sliceStr = `:function\\(a,b\\)\\{return a\\.slice\\(b\\)\\}`;
const spliceStr = `:function\\(a,b\\)\\{a\\.splice\\(0,b\\)\\}`;
const swapStr = `:function\\(a,b\\)\\{var c=a\\[0\\];a\\[0\\]=a\\[b(?:%a\\.length)?\\];a\\[b(?:%a\\.length)?\\]=c(?:;return a)?\\}`;

const actionsObjRegex = new RegExp(
    `var (${jsVarStr})=\\{((?:(?:${jsKeyStr}${reverseStr}|${jsKeyStr}${sliceStr}|${jsKeyStr}${spliceStr}|${jsKeyStr}${swapStr}),?\\r?\\n?)+)\\};`
);

const actionsFuncRegex = new RegExp(
    `function(?: ${jsVarStr})?\\(a\\)\\{a=a\\.split\\((?:''|"")\\);\\s*((?:(?:a=)?${jsVarStr}${jsPropStr}\\(a,\\d+\\);)+)return a\\.join\\((?:''|"")\\)\\}`
);

const reverseRegex = new RegExp(`(?:^|,)(${jsKeyStr})${reverseStr}`);
const sliceRegex = new RegExp(`(?:^|,)(${jsKeyStr})${sliceStr}`);
const spliceRegex = new RegExp(`(?:^|,)(${jsKeyStr})${spliceStr}`);
const swapRegex = new RegExp(`(?:^|,)(${jsKeyStr})${swapStr}`);

// Decipherer handles signature and n-parameter challenges
export class Decipherer {
    private playerCode: string;
    private playerURL: string;
    private sigTokens: string[];
    private nSolver: NSolver | null;

    constructor(playerCode: string, playerURL: string = "") {
        this.playerCode = playerCode;
        this.playerURL = playerURL;
        this.sigTokens = [];
        this.nSolver = null;
    }

    // Initialize extracts signature tokens and creates n-solver
    async Initialize(): Promise<void> {
        this.extractSignatureTokens();

        // Initialize n-solver
        try {
            this.nSolver = new NSolver(this.playerCode);
        } catch {
            // N-solver is optional, some videos don't need it
            this.nSolver = null;
        }
    }

    // GetSignatureTimestamp returns the signature timestamp from the player code
    GetSignatureTimestamp(): number {
        return GetSignatureTimestamp(this.playerCode);
    }

    // SolveNChallenge solves the n-parameter challenge using the JS runtime
    SolveNChallenge(n: string): string {
        if (!this.nSolver) {
            return n;
        }
        return this.nSolver.Solve(n);
    }

    // DecipherURL deciphers a stream URL by solving signature and n-parameter challenges
    DecipherURL(streamURL: string): string {
        const parsed = new URL(streamURL);
        const query = parsed.searchParams;

        // Check for signature cipher in query
        const sig = query.get("s");
        if (sig) {
            // Decipher signature
            const decipheredSig = this.decipherSignature(sig);

            // Get signature parameter name (usually "sig" or "signature")
            let sp = query.get("sp");
            if (!sp) {
                sp = "signature";
            }

            query.set(sp, decipheredSig);
            query.delete("s");
            query.delete("sp");
        }

        // Handle n-parameter for throttle bypass
        const n = query.get("n");
        if (n) {
            const newN = this.SolveNChallenge(n);
            if (newN) {
                query.set("n", newN);
            }
        }

        return parsed.toString();
    }

    // DecipherSignature deciphers a signature using the extracted tokens
    DecipherSignature(sig: string): string {
        return this.decipherSignature(sig);
    }

    // decipherSignature applies the signature transformation
    private decipherSignature(sig: string): string {
        let arr = sig.split("");

        for (const token of this.sigTokens) {
            if (token.length < 1) {
                continue;
            }

            switch (token[0]) {
                case "r":
                    // Reverse
                    arr = arr.reverse();
                    break;

                case "s": {
                    // Slice
                    const pos = parseInt(token.slice(1), 10);
                    if (!isNaN(pos) && pos < arr.length) {
                        arr = arr.slice(pos);
                    }
                    break;
                }

                case "p": {
                    // Splice
                    const pos = parseInt(token.slice(1), 10);
                    if (!isNaN(pos) && pos < arr.length) {
                        arr = arr.slice(pos);
                    }
                    break;
                }

                case "w": {
                    // Swap
                    const pos = parseInt(token.slice(1), 10);
                    if (!isNaN(pos)) {
                        const swapPos = pos % arr.length;
                        [arr[0], arr[swapPos]] = [arr[swapPos], arr[0]];
                    }
                    break;
                }
            }
        }

        return arr.join("");
    }

    // extractSignatureTokens extracts signature transformation tokens from player code
    private extractSignatureTokens(): void {
        const objects = this.playerCode.match(actionsObjRegex);
        const functions = this.playerCode.match(actionsFuncRegex);

        if (!objects || objects.length < 3 || !functions || functions.length < 2) {
            // Try alternative extraction methods
            this.extractSignatureTokensAlt();
            return;
        }

        const obj = objects[1].replace(/\$/g, "\\$");
        const objBody = objects[2].replace(/\$/g, "\\$");
        const funcBody = functions[1].replace(/\$/g, "\\$");

        const reverseKey = extractKey(reverseRegex, objBody);
        const sliceKey = extractKey(sliceRegex, objBody);
        const spliceKey = extractKey(spliceRegex, objBody);
        const swapKey = extractKey(swapRegex, objBody);

        const keys = `(${reverseKey}|${sliceKey}|${spliceKey}|${swapKey})`;
        const tokenizeRegex = new RegExp(
            `(?:a=)?${obj}(?:\\.${keys}|\\[(?:'${keys}'|"${keys}")\\])\\(a,(\\d+)\\)`,
            "g"
        );

        let match;
        while ((match = tokenizeRegex.exec(funcBody)) !== null) {
            if (match.length < 5) {
                continue;
            }

            let key = match[1];
            if (!key) {
                key = match[2];
            }
            if (!key) {
                key = match[3];
            }

            switch (key) {
                case reverseKey:
                    this.sigTokens.push("r");
                    break;
                case sliceKey:
                    this.sigTokens.push("s" + match[4]);
                    break;
                case spliceKey:
                    this.sigTokens.push("p" + match[4]);
                    break;
                case swapKey:
                    this.sigTokens.push("w" + match[4]);
                    break;
            }
        }
    }

    // extractSignatureTokensAlt tries alternative patterns for signature extraction
    private extractSignatureTokensAlt(): void {
        // Alternative signature function pattern
        const altPattern = /\b[cs]\s*&&\s*[adf]\.set\([^,]+\s*,\s*encodeURIComponent\(([a-zA-Z0-9$]+)\(/;
        const match = this.playerCode.match(altPattern);

        if (!match || match.length < 2) {
            // If no signature function found, the URL may not need deciphering
            return;
        }

        // Further extraction would go here
    }
}

// extractKey extracts a key from the object body using the given regex
function extractKey(re: RegExp, body: string): string {
    const match = body.match(re);
    if (!match || match.length < 2) {
        return "";
    }

    let key = match[1];
    // Remove quotes if present
    key = key.replace(/^["']|["']$/g, "");
    return key;
}

// GetSignatureTimestamp extracts the signature timestamp from player code
export function GetSignatureTimestamp(playerCode: string): number {
    const patterns = [/(?:signatureTimestamp|sts)\s*:\s*(\d{5})/, /"STS"\s*:\s*(\d{5})/];

    for (const pattern of patterns) {
        const match = playerCode.match(pattern);
        if (match && match[1]) {
            const sts = parseInt(match[1], 10);
            if (!isNaN(sts)) {
                return sts;
            }
        }
    }

    return 0;
}

// ExtractPlayerID extracts the player ID from a player URL
export function ExtractPlayerID(playerURL: string): string {
    const patterns = [/\/s\/player\/([a-zA-Z0-9_-]{8,})\//, /\/([a-zA-Z0-9_-]{8,})\/player/, /\b(vfl[a-zA-Z0-9_-]+)\b/];

    for (const pattern of patterns) {
        const match = playerURL.match(pattern);
        if (match && match[1]) {
            return match[1];
        }
    }

    return "";
}

// NewDecipherer creates a new Decipherer from player code
export async function NewDecipherer(playerCode: string, playerURL: string = ""): Promise<Decipherer> {
    const d = new Decipherer(playerCode, playerURL);
    await d.Initialize();
    return d;
}
